File: Example/javaapp/src/main/java/com/stripe/example/javaapp/fragment/location/LocationListAdapter.java
package com.stripe.example.javaapp.fragment.location;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.stripe.example.javaapp.R;
import com.stripe.stripeterminal.external.models.Location;

/**
 * Adapts a list of locations for a RecyclerView with a header and loading view.
 */
public class LocationListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private final LayoutInflater inflater;
    private final LocationSelectionController locationSelectionController;
    private LocationListState locationListState = null;

    public LocationListAdapter(
        LayoutInflater inflater,
        LocationSelectionController locationSelectionController
    ) {
        super();
        this.inflater = inflater;
        this.locationSelectionController = locationSelectionController;
    }

    public void setLocationListState(LocationListState state) {
        locationListState = state;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == R.layout.list_item_card) {
            return new LocationListViewHolder(
                    inflater.inflate(viewType, parent, false),
                    locationSelectionController
            );
        } else if (viewType == R.layout.list_item_progress) {
            return new ProgressViewHolder(
                    inflater.inflate(viewType, parent, false)
            );
        } else if (viewType == R.layout.list_item_header) {
            return new HeaderViewHolder(
                    inflater.inflate(viewType, parent, false)
            );
        } else {
            throw new IllegalStateException("Unknown View Type: " + viewType);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof LocationListViewHolder) {
            Location location = locationListState.getLocations().get(position - 1);
            ((LocationListViewHolder) holder).bind(location);
        } else if (holder instanceof HeaderViewHolder) {
            int size = locationListState == null ? 0 : locationListState.getLocations().size();
            boolean hasMore = locationListState != null && locationListState.hasMore();
            ((HeaderViewHolder) holder).bind(size, hasMore);
        }
    }

    @Override
    public int getItemCount() {
        if (locationListState == null) return 0;

        int header = locationListState.headerVisible() ? 1 : 0;
        int loading = locationListState.isLoading() ? 1 : 0;
        int locations = locationListState.getLocations().size();

        return header + locations + loading;
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0 && locationListState.headerVisible()) {
            return R.layout.list_item_header;
        } else if (position == getItemCount() - 1 && locationListState.isLoading()) {
            return R.layout.list_item_progress;
        } else {
            return R.layout.list_item_card;
        }
    }
}

File: Example/kotlinapp/src/main/java/com/stripe/example/fragment/location/LocationListAdapter.kt
package com.stripe.example.fragment.location

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.stripe.example.R
import java.lang.IllegalStateException

/**
 * Adapts a list of locations for a RecyclerView with a header and loading view.
 */
class LocationListAdapter(
    private val inflater: LayoutInflater,
    private val locationSelectionController: LocationSelectionController,
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    var locationListState: LocationListState? = null
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            R.layout.list_item_card -> LocationListViewHolder(
                parent = inflater.inflate(viewType, parent, false),
                locationSelectionController = locationSelectionController
            )
            R.layout.list_item_progress -> ProgressViewHolder(
                parent = inflater.inflate(viewType, parent, false)
            )
            R.layout.list_item_header -> HeaderViewHolder(
                parent = inflater.inflate(viewType, parent, false)
            )
            else -> throw IllegalStateException("Unknown View Type: $viewType")
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder) {
            is LocationListViewHolder ->
                locationListState?.locations
                    ?.get(position - 1)
                    ?.also { holder.bind(it) }
            is HeaderViewHolder -> holder.bind(locationListState?.locations?.size ?: 0, locationListState?.hasMore ?: false)
        }
    }

    override fun getItemCount(): Int {
        val header = if (locationListState?.headerVisible == true) 1 else 0
        val loading = if (locationListState?.isLoading == true) 1 else 0
        val locations = locationListState?.locations?.size ?: 0

        return header + locations + loading
    }

    override fun getItemViewType(position: Int): Int = when {
        position == 0 && locationListState?.headerVisible == true -> R.layout.list_item_header
        position == itemCount - 1 && locationListState?.isLoading == true -> R.layout.list_item_progress
        else -> R.layout.list_item_card
    }
}
